assets/src/blocks/Submenu/SubmenuEditor.js

```diff
@@ -0,0 +1,123 @@
+import { Fragment } from '@wordpress/element';
+import { Button, PanelBody } from '@wordpress/components';
+import { SubmenuLevel } from './SubmenuLevel';
+import { SubmenuItems } from './SubmenuItems';
+import { InspectorControls } from '@wordpress/block-editor';
+import { getSubmenuStyle } from './getSubmenuStyle';
+import { makeHierarchical } from './makeHierarchical';
+import { getHeadingsFromBlocks} from './getHeadingsFromBlocks';
+import { useSelect } from '@wordpress/data';
+
+const { __ } = wp.i18n;
+const { RichText } = wp.blockEditor;
+const renderEdit = (attributes, setAttributes) => {
+  function addLevel() {
+    const [previousLastLevel] = attributes.levels.slice(-1);
+    const newLevel = previousLastLevel.heading + 1;
+    setAttributes({ levels: attributes.levels.concat({ heading: newLevel, link: false, style: 'none' }) });
+  }
+
+  function onHeadingChange(index, value) {
+    let levels = JSON.parse(JSON.stringify(attributes.levels));
```

Comment:
Not a blocker: This should be safe given the content of `levels` but I personally prefer using LoDash's `deepClone` because it's safer for different types and faster, but also is more expressive when reading the code for someone that is not familiar with this technique.

Reply:
Yes LoDash `deepClone` is better for this. But I actually didn't touch `SubmenuLevel` yet, it's still a class. Maybe it can be written in a way that doesn't require the `index` lookup, and maybe also doesn't need a deep clone. Then we can avoid adding a dependency on LoDash here.

Reply:
I made a function that has the same output as the JSON switcheroo has, and is a few times faster than it. It's not as "feature" complete as Lodash, in the sense that it doesn't preserve non-built in classes, but so doesn't the JSON approach we used :shrug: There are also some exotic or deprecated JavaScript peculiarities that don't behave the same way as in Lodash, but we don't use them and we probably never should :smile:

Reply:
Did a quick benchmark
```
/usr/bin/node /home/pieter/deepclone/bm.js
JSON bytes:  18
matches JSON approach
JSON-switcheroo   :  x 1,306,315 ops/sec ±1.03% (85 runs sampled)
lodash            :  x 1,400,846 ops/sec ±1.91% (87 runs sampled)
Custom check refs :  x 5,385,561 ops/sec ±3.58% (83 runs sampled)
Fastest is Custom check refs : 
Fastest is [object Object]
JSON bytes:  2824
matches JSON approach
JSON-switcheroo   :  x 41,676 ops/sec ±0.77% (88 runs sampled)
lodash            :  x 40,323 ops/sec ±1.27% (85 runs sampled)
Custom check refs :  x 91,897 ops/sec ±1.79% (90 runs sampled)
Fastest is Custom check refs : 
Fastest is [object Object]
JSON bytes:  7374
matches JSON approach
JSON-switcheroo   :  x 13,925 ops/sec ±1.13% (91 runs sampled)
lodash            :  x 17,017 ops/sec ±1.41% (83 runs sampled)
Custom check refs :  x 40,898 ops/sec ±3.23% (83 runs sampled)
Fastest is Custom check refs : 
Fastest is [object Object]
JSON bytes:  122662
matches JSON approach
JSON-switcheroo   :  x 762 ops/sec ±1.33% (85 runs sampled)
lodash            :  x 887 ops/sec ±1.30% (86 runs sampled)
Custom check refs :  x 1,873 ops/sec ±1.22% (86 runs sampled)
Fastest is Custom check refs : 
Fastest is [object Object]
JSON bytes:  642680
matches JSON approach
JSON-switcheroo   :  x 176 ops/sec ±2.45% (69 runs sampled)
lodash            :  x 295 ops/sec ±1.61% (82 runs sampled)
Custom check refs :  x 243 ops/sec ±1.09% (81 runs sampled)
Fastest is lodash            : 
Fastest is [object Object]

Process finished with exit code 0

```